src/goPackages.ts

```diff
@@ -19,37 +12,31 @@ export function isGoListComplete(): boolean {
 export function goListAll(): Promise<Map<string, string>> {
 	let goRuntimePath = getGoRuntimePath();
 
+	// TODO prompt for missing tools, attach cmd.on('error')
 	if (!goRuntimePath) {
 		vscode.window.showInformationMessage('Cannot find "go" binary. Update PATH or GOROOT appropriately');
 		return Promise.resolve(null);
 	}
 
-	if (goListAllPromise) {
-		return goListAllPromise;
-	}
-
-	goListAllPromise = new Promise<Map<string, string>>((resolve, reject) => {
-		// Use `{env: {}}` to make the execution faster. Include GOPATH to account if custom work space exists.
-		const env: any = getToolsEnvVars();
-
-		const cmd = cp.spawn(goRuntimePath, ['list', '-f', '{{.Name}};{{.ImportPath}}', 'all'], { env: env });
+	return new Promise<Map<string, string>>((resolve, reject) => {
+		const cmd = cp.spawn(getBinPath('gopkgs'), ['-short=false'], { env: getToolsEnvVars() });
 		const chunks = [];
 		cmd.stdout.on('data', (d) => {
 			chunks.push(d);
 		});
 
 		cmd.on('close', (status) => {
-			chunks.join('').split('\n').forEach(pkgDetail => {
-				if (!pkgDetail || !pkgDetail.trim() || pkgDetail.indexOf(';') === -1) return;
-				let [pkgName, pkgPath] = pkgDetail.trim().split(';');
-				allPkgs.set(pkgPath, pkgName);
+			// TODO do we need to cache?
+			let pkgs = new Map<string, string>();
+			chunks.join('').split('\n').forEach((pkgPath) => {
+				if (!pkgPath) return;
+				const lastIndex = pkgPath.lastIndexOf('/');
```

Comment:
The main reason I used `go list` instead of `gopkgs` initially was to get the package name and not rely on the import path of the package to determine the package name like you are doing here.

Because package names need not necessarily be the last part of the package path. See https://github.com/Microsoft/vscode-go/issues/647

And we use the package names to provide completion for unimported packages.

Reply:
I see... make sense, because it completion, not import package

Reply:
Yes, and later on I added the `Go: Browse Packages` feature because then I had all the packages in hand, so it seemed like a good feature to add

Reply:
We can strip off the version part from "pkgname.v1" , it works when people always using standard way on naming package related to the package path, but that doesn't guarantee

Reply:
That's exactly what I did in https://github.com/Microsoft/vscode-go/commit/cb0d2bc29a0c9ee1929c1a89781ef8c740447d92 :)

Reply:
[cb0d2bc](https://github.com/Microsoft/vscode-go/commit/cb0d2bc29a0c9ee1929c1a89781ef8c740447d92) no longer in master?

Hare are some example, can't guarantee consistent pattern:
- speech;google.golang.org/api/speech/v1beta1
- spectrum;google.golang.org/api/spectrum/v1explorer
- speech;google.golang.org/api/speech/v1
- prediction;google.golang.org/api/prediction/v1.2
- tilde;gopkg.in/mattes/go-expand-tilde.v1

So using `go list -f '{{.Name}};{{.ImportPath}}' all` is the proper way, it just need to make it faster.

Case
1. Autocomplete un-imported will suggest only packages on cache, no way to suggest new package since it need invalidation. Need to listen some events to invalidates, haven't figure out which one.
2. Go: Browse Packages use `https://github.com/tpng/gopkgs` which looks in `$GOPATH/pkg`, new package will not exists in `$GOPATH/pkg`, so browse package will not help. Except we change it, use `https://github.com/haya14busa/gopkgs` which search in   `$GOPATH/src`. This is workaround, but still completion still stale.

Btw, why not change to `https://github.com/haya14busa/gopkgs` for Browse Package? It seems good improvement, solve one problem.

Reply:
Yes, I replaced https://github.com/Microsoft/vscode-go/commit/cb0d2bc29a0c9ee1929c1a89781ef8c740447d92 with using `go list -f '{{.Name}};{{.ImportPath}}' all`

I am fine with auto complete using cache as it does today. 

I didnt use `https://github.com/tpng/gopkgs` for the `Browse Packages` feature as it will not have the "main" packages as in the ones that are executable

I am fine with using `https://github.com/haya14busa/gopkgs` for just the Browse Packages feature as well